src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java

```diff
@@ -126,7 +130,31 @@ public final void checkout(Run<?,?> run, FilePath workspace, TaskListener listen
                 }
                 }
             }
-            scm.checkout(run, launcher, workspace, listener, changelogFile, baseline);
+
+            for (int retryCount = Jenkins.get().getScmCheckoutRetryCount(); retryCount >= 0; retryCount--) {
+                try {
+                    scm.checkout(run, launcher, workspace, listener, changelogFile, baseline);
+                    break;
+                } catch (AbortException e) {
+                    // abort exception might have a null message.
+                    // If so, just skip echoing it.
+                    if (e.getMessage() != null) {
+                        listener.error(e.getMessage());
+                    }
+                } catch (InterruptedIOException e) {
+                    throw e;
+                } catch (Exception e) {
+                    // checkout error not yet reported
+                    Functions.printStackTrace(e, listener.error("Checkout failed"));
+                }
+
+                if (retryCount == 0)   // all attempts failed
+                    throw new AbortException("Maximum checkout retry attempts reached, aborting");
+
+                listener.getLogger().println("Retrying after 10 seconds");
+                Thread.sleep(10000);
```

Comment:
This does not look right to me, should probably do something like what the sleep step does instead of blocking the thread here. Though I am not sure atm how a `SynchronousNonBlockingStepExecution` is supposed to behave in these situations.

Reply:
Well, a background thread will sleep. Not terrible, under the assumption this code path is rarely used.

Reply:
Shall I keep this change? Or, do I need to rewrite the whole StepExecutionImpl?

Reply:
Probably good enough as is. This step used `SynchronousNonBlockingStepExecution` (or the equivalent) from the beginning since the `SCM` interface in Jenkins does not support durable execution, so unlike

```groovy
sh 'git clone https://github.com/myorg/myrepo'
```

a slow checkout cannot survive a controller restart. As of https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/203 this matters less anyway since you can use the new retry system with the `nonresumable` condition to handle that case.